### PR TITLE
add mysqli support

### PR DESCRIPTION
**Pull request by xonze**

### CODE REVIEW

**Review by xonze — COMMENTED** [review]
ok

**Comment by heyanlong** [comment]
@xonze Thanks for the code, take a look at the review content

**Review by xonze — COMMENTED** [review]
sky_plugin_error support php8 zend_error_cb hook

**Review by xonze — COMMENTED** [review]
ok

**Comment by xonze** [comment]
@heyanlong Please  Review this code

**Review by xonze — COMMENTED** [review]
That's  all  the code  for  this  pull  request

**Comment by xonze** [comment]
**skywalking_log() and skywalking_tag() example:**
>   <?php
  class Skywalking
  {

    public function log($msg)
    {
        if ($msg != "" && function_exists("skywalking_log")) {
            skywalking_log($msg);
        }
        return $this;
    }

    public function tag($key, $value)
    {
        if ($key != "" && $value != "" && function_exists("skywalking_tag")) {
            skywalking_tag($key, $value);
        }
        return $this;
    }

    public function getTraceId()
    {
        if (function_exists("skywalking_trace_id")) {
            return skywalking_trace_id();
        }

        return "";
    }
}
$sw = new Skywalking();
$sw->tag("foo", "bar");
$sw->tag("langue","en")
   ->tag("zh", "中文")
   ->tag("项目", "测试项目")
   ->log(
       "自定义日志函数【skywalking_log(string msg)】\n" .
       "自定义标签函数【skywalking_tag(string \$key, string \$value)】\n" .
       __FILE__ . '(' . __LINE__ . ")\n" .
       print_r($sw, true)
    );
var_dump($sw->getTraceId());

File: config.m4 [hunk 1]
```diff
@@ -144,11 +144,14 @@ if test "$PHP_SKYWALKING" != "no"; then
       src/manager.cc \
       src/segment.cc \
       src/segment_reference.cc \
+      src/span_log.cc \
       src/sky_execute.cc \
       src/sky_grpc.cc \
       src/sky_module.cc \
+      src/sky_mysqli.cc \
```
Comment (heyanlong) [inline config.m4:151]:
```suggestion
      src/sky_plugin_mysqli.cc \
```

File: package.xml [hunk 2]
```diff
@@ -81,6 +83,8 @@
 			<file role="src" name="src/sky_grpc.h"/>
 			<file role="src" name="src/sky_module.cc"/>
 			<file role="src" name="src/sky_module.h"/>
+			<file role="src" name="src/sky_mysqli.cc"/>
+			<file role="src" name="src/sky_mysqli.h"/>
```
Comment (heyanlong) [inline package.xml:87]:
```suggestion
			<file role="src" name="src/sky_plugin_mysqli.cc"/>
			<file role="src" name="src/sky_plugin_mysqli.h"/>
```

File: skywalking.cc [hunk 3]
```diff
@@ -87,6 +91,47 @@ PHP_FUNCTION (skywalking_trace_id) {
...
+        Z_PARAM_OPTIONAL
+        Z_PARAM_BOOL(is_error)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (!SKYWALKING_G(enable) || SKYWALKING_G(segment) == nullptr) {
+        return;
+    }
+
+    auto *segment = static_cast<Segment *>(SKYWALKING_G(segment));
+    segment->addErrorLog("CUSTOM_LOG", log->val, is_error);
```
Comment (heyanlong) [inline skywalking.cc:111]: skywalking_log("key", "value") ?

File: src/sky_execute.cc [hunk 4]
```diff
@@ -24,6 +24,7 @@
 #include "sky_plugin_hyperf_guzzle.h"
 #include "sky_plugin_swoole_curl.h"
 #include "sky_pdo.h"
+#include "sky_mysqli.h"
```
Comment (heyanlong) [inline src/sky_execute.cc:27]:
```suggestion
#include "sky_plugin_mysqli.h"
```

File: src/sky_plugin_swoole_curl.cc [hunk 5]
```diff
@@ -79,7 +79,7 @@ void sky_plugin_swoole_curl(zend_execute_data *execute_data, const std::string &
 
             if (!Z_ISUNDEF_P(http_code) && Z_TYPE_P(http_code) == IS_LONG) {
                 span->addTag("status_code", std::to_string(Z_LVAL_P(http_code)));
-                if (Z_LVAL_P(http_code) >= 400) {
+                if (Z_LVAL_P(http_code) >= 400 || Z_LVAL_P(http_code) == 0) {
```
Comment (heyanlong) [inline src/sky_plugin_swoole_curl.cc:82]: Why ?

**Review by heyanlong — COMMENTED** [review]
Proposal to continue after modification

File: src/sky_plugin_curl.cc [hunk 6]
```diff
@@ -218,7 +218,7 @@ void sky_curl_exec_handler(INTERNAL_FUNCTION_PARAMETERS) {
         zval *response_http_code;
         response_http_code = zend_hash_str_find(Z_ARRVAL(url_response), ZEND_STRL("http_code"));
         span->addTag("status_code", std::to_string(Z_LVAL_P(response_http_code)));
-        if (Z_LVAL_P(response_http_code) >= 400) {
+        if (Z_LVAL_P(response_http_code) == 0 || Z_LVAL_P(response_http_code) >= 400) {
```
[the remark this answers is not in the document]
Comment (xonze) [reply src/sky_plugin_curl.cc]: When  domian can't resovle, response_http_code is 0

File: src/sky_plugin_swoole_curl.cc (hunk 5 above)
Comment (xonze) [reply src/sky_plugin_swoole_curl.cc:82]: When domian can't resovle, http_code is 0

Comment (heyanlong) [reply src/sky_plugin_swoole_curl.cc:82]: According to other SDK implementations, all achieve 0, I expect to be consistent

**Review by xonze — COMMENTED** [review]
ok

**Review by xonze — COMMENTED** [review]
ok

**Comment by heyanlong** [comment]
Is there an E2E test?

File: src/sky_execute.cc [hunk 7]
```diff
@@ -111,6 +112,13 @@ void sky_execute_internal(zend_execute_data *execute_data, zval *return_value) {
         char *class_name = is_class ? ZSTR_VAL(fn->common.scope->name) : nullptr;
         char *function_name = fn->common.function_name != nullptr ? ZSTR_VAL(fn->common.function_name) : nullptr;
 
+        if (function_name != nullptr && 
+            SKYWALKING_G(error_handler_enable) && 
+            strcmp(function_name, "set_error_handler") == 0) {
+            php_error(E_NOTICE, "'skywalking.error_handler_enable' current congfig is 'enable', The 'set_error_handler' function will not work, The PHP error log has been handle and reporting to skywalking server.");
+            return;
+        }
+
```
Comment (xonze) [inline src/sky_execute.cc:121]: @heyanlong Is the intercept set_error_handler  function method appropriate？

**Comment by xonze** [comment]
> Is there an E2E test?

I don't  know how  to do it.

**Comment by heyanlong** [comment]
https://github.com/SkyAPM/SkyAPM-php-sdk/blob/master/e2e/tests/redis.php

File: e2e/index.php [hunk 8]
```diff
@@ -6,4 +6,8 @@
 $ch = curl_init("https://api.github.com/repos");
 curl_exec($ch);
 
-testRedis();
\ No newline at end of file
+testRedis();
+
+testAddTag();
```
Comment (xonze) [inline e2e/index.php:11]: Is it wrong to write like this?

Comment (heyanlong) [reply e2e/index.php:11]: yes.

**Comment by heyanlong** [comment]
Can I invite you to the WeChat development group?

**Comment by xonze** [comment]
> Can I invite you to the WeChat development group?

yes，ID is： wenxiangli

**Comment by heyanlong** [comment]
This branch has conflicts that must be resolved
Conflicting files
src/sky_execute.cc
src/sky_module.cc
src/sky_utils.cc

**Review by xonze — COMMENTED** [review]
conflicts all resolved

**Review by heyanlong — APPROVED** [review]

**Pull request merged**